File: tests/fixtures/cover-behaviour-gate/behaviour.ts
// Fixture for tests/unit/test-cover-behaviour-gate.mjs: every test here asserts behaviour.
// A commented-out read must not count: readFileSync('src/pricing.ts')
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, it, test, vi } from 'vitest';
import { applyDiscount, parseAmount, sendReceipt } from '../src/pricing';

const cases = JSON.parse(readFileSync(join(__dirname, 'fixtures', 'cases.json'), 'utf8'));

describe('applyDiscount', () => {
  it('returns the discounted total', () => {
    expect(applyDiscount(100, 0.2)).toBe(80);
  });

  it('rejects a negative rate with a RangeError', () => {
    expect(() => applyDiscount(100, -1)).toThrow(RangeError);
  });

  it('sends the receipt and returns its id', async () => {
    const mailer = { send: vi.fn().mockResolvedValue({ id: 'r1' }) };
    const receipt = await sendReceipt(mailer, 42);
    expect(mailer.send).toHaveBeenCalledWith(expect.objectContaining({ orderId: 42 }));
    expect(receipt.id).toBe('r1');
  });

  it('ignores "expect(" and a regex with a paren /\\(/ in strings', () => {
    const pattern = /\(total\)/;
    expect(pattern.test('(total)')).toBe(true);
  });

  it.each(cases)('parses %s', (input, expected) => {
    expect(parseAmount(input)).toEqual(expected);
  });

  it('does not throw on an empty cart', () => {
    expect(() => applyDiscount(0, 0)).not.toThrow();
  });

  test('parses with a node assert', () => {
    assert.strictEqual(parseAmount('1.50'), 150);
  });
});

File: tests/fixtures/cover-behaviour-gate/behaviour.py
# Fixture for tests/unit/test-cover-behaviour-gate.mjs: every test here asserts behaviour.
import json
import unittest
from pathlib import Path
from unittest.mock import MagicMock

import pytest
from app.pricing import apply_discount, parse_amount, send_receipt

CASES = json.loads((Path(__file__).parent / "fixtures" / "cases.json").read_text())


def test_returns_discounted_total():
    assert apply_discount(100, 0.2) == 80


def test_negative_rate_raises():
    with pytest.raises(ValueError, match="rate"):
        apply_discount(100, -1)


def test_sends_receipt_and_returns_id():
    mailer = MagicMock()
    mailer.send.return_value = {"id": "r1"}
    receipt = send_receipt(mailer, 42)
    mailer.send.assert_called_once()
    assert receipt["id"] == "r1"


def test_empty_cart():
    """Not raising on an empty cart is the contract: apply_discount never raises for zero totals."""
    apply_discount(0, 0)


class TestParse(unittest.TestCase):
    def test_parses_cents(self):
        self.assertEqual(parse_amount("1.50"), 150)

    def test_comment_mentions_assert_called_but_asserts_value(self):
        # mock.assert_called() would be rule (a); this test checks the value instead
        self.assertIsNone(parse_amount(""))
